File: dnd.py
# dnd - pennbauman.com
#   URL: https://github.com/pennbauman/pennbauman.com
#   License: GPLv3.0
#   Author:
#     Penn Bauman (pennbauman@protonmail.com)
from flask import Blueprint, render_template, abort

dnd_app = Blueprint('dnd', __name__,
                        template_folder='templates')

# Unique pages
@dnd_app.route("/dnd")
def dnd():
    return render_template("dnd/index.html")

@dnd_app.route("/dnd/dice")
def dice():
    return render_template("dnd/dice.html")

@dnd_app.route("/dnd/stats")
def stats():
    return render_template("dnd/stats.html",
            stats = ["Str", "Dex", "Con", "Int", "Wis", "Cha"])

# Spell pages
@dnd_app.route("/dnd/spells")
def spell_index():
    return render_template("dnd/spells.html")
@dnd_app.route("/dnd/spells/<school>")
def spell_schools(school):
    try:
        return render_template("dnd/spells/" + school + ".html")
    except:
        abort(404)


# Monster pages
@dnd_app.route("/dnd/monsters")
def monster_index():
    return render_template("dnd/monsters.html")
@dnd_app.route("/dnd/monsters/<category>")
def monster_categories(category):
    try:
        return render_template("dnd/monsters/" + category + ".html")
    except:
        abort(404)

File: src/apps/dnd.rs
// dnd pages - pennbauman.com
//   URL: https://github.com/pennbauman/pennbauman.com
//   License: GPLv3.0
//   Author:
//     Penn Bauman (pennbauman@protonmail.com)
use tide::{Request, Response};
use tide_fluent_routes::prelude::*;
use tide_tera::prelude::*;
use crate::State;
use crate::tera_page;

pub fn routes(routes: RouteSegment<State>) -> RouteSegment<State> {
    routes.all(tera_page!("dnd/index.html"))
        .at("", |route| route.all(tera_page!("dnd/index.html")))
        .at("stats", |route| route.all(tera_page!("dnd/stats.html")))
        .at("dice", |route| route.all(tera_page!("dnd/dice.html")))
        .at("spells", |route| route.all(tera_page!("dnd/spells.html")))
        .at("monsters", |route| route.all(tera_page!("dnd/monsters.html")))

        // Spells
        .at("spells/:school", |route| route.all(|req: Request<State>| async move {
            let school: &str = req.param("school")?;
            let tera = &req.state().tera;
            let page = tera.render_response(&format!("dnd/spells/{}.html", school), &context!());
            match page {
                Ok(p) => Ok(p),
                Err(_) => Ok(Response::new(404)),
            }
        }))

        // Monsters
        .at("monsters/:type", |route| route.all(|req: Request<State>| async move {
            let school: &str = req.param("type")?;
            let tera = &req.state().tera;
            let page = tera.render_response(&format!("dnd/monsters/{}.html", school), &context!());
            match page {
                Ok(p) => Ok(p),
                Err(_) => Ok(Response::new(404)),
            }
        }))
}
